File: quotation/core/xml_reader.py
"""eConfig XML 파서."""
from __future__ import annotations

import codecs
import re
from pathlib import Path

from lxml import etree

from . import integrated, modes
from .models import Group, LineItem, Quotation, SubLineItem
from .money import parse_amount
from .naming import item_key, safe_sheet_name, sheet_name, unique_sheet_names

# --- 원본 프로그램의 XPath (변경 금지) ---------------------------------------
XP_CFDATA = "./CFData"
XP_LINE_ITEM = ".//ProductLineItem"
XP_SUB_LINE_ITEM = "./ProductSubLineItem"

XP_LINE_NUMBER = "./ProductLineNumber"
XP_SIU = "./CPUSIUvalue"
XP_TXN_TYPE = "./TransactionType"
XP_GROUP_ID = "./ProprietaryGroupIdentifier"
XP_QUANTITY = "./Quantity"
XP_DESC = "./ProductIdentification/PartnerProductIdentification/ProductDescription"
XP_PRODUCT_NAME = "./ProductIdentification/PartnerProductIdentification/ProductName"
XP_TYPE_CODE = "./ProductIdentification/PartnerProductIdentification/ProductTypeCode"
XP_PART_NO = ("./ProductIdentification/PartnerProductIdentification"
              "/ProprietaryProductIdentifier")
XP_AMOUNT = "./UnitListPrice/FinancialAmount/MonetaryAmount"


class QuotationXmlError(Exception):
    """XML 이 견적서 생성 요건을 만족하지 않을 때. 메시지는 원본 프로그램과 동일하다."""


def _text(el, path: str) -> str:
    node = el.find(path)
    if node is None or node.text is None:
        return ""
    return node.text.strip()


def _int(el, path: str, default: int = 1) -> int:
    return int(_text(el, path) or default)


def _parse_sub(el) -> SubLineItem:
    return SubLineItem(
        txn_type=_text(el, XP_TXN_TYPE),
        quantity=_int(el, XP_QUANTITY),
        part_number=_text(el, XP_PART_NO),
        description=_text(el, XP_DESC),
        unit_price=parse_amount(_text(el, XP_AMOUNT) or None),
    )


def _parse_line(el) -> LineItem:
    return LineItem(
        line_number=_text(el, XP_LINE_NUMBER),
        txn_type=_text(el, XP_TXN_TYPE),
        group_id=_text(el, XP_GROUP_ID),
        quantity=_int(el, XP_QUANTITY),
        part_number=_text(el, XP_PART_NO),
        description=_text(el, XP_DESC),
        product_type=_text(el, XP_TYPE_CODE),
        product_name=_text(el, XP_PRODUCT_NAME),
        unit_price=parse_amount(_text(el, XP_AMOUNT) or None),
        subs=tuple(_parse_sub(s) for s in el.findall(XP_SUB_LINE_ITEM)),
        siu=_int(el, XP_SIU, default=0),
    )


def _reference_names(items: list[LineItem]) -> list[str]:
    """증설 견적의 장비 이름. 기존(BASE)/증설후(PROPOSED) 구성의 본체 라인에서 딴다.

    증설 라인(UPGRADE)의 Description 은 '9080 Model HEU' 처럼 장비 이름이 없다.
    골든은 시트를 'SERVER 1' 로 부르는데, 그 이름은 BASE 구성의 본체 라인
    'Server 1:Server 1:IBM Power E1080' 에서 온다.
    """
    seen: set[str] = set()
    names: list[str] = []
    for it in items:
        if it.is_reference and it.siu == 1:
            key = item_key(it.description)
            if key not in seen:
                seen.add(key)
                names.append(key)
    return names


def _build_groups(items: list[LineItem], reference_names: list[str],
                  mode: str = modes.DEFAULT) -> tuple[Group, ...]:
    """ProprietaryGroupIdentifier 로 묶는다. 문서 등장 순서를 유지한다.

    증설 견적일 때만 두 가지 보정이 붙는다.
      - 본체 라인(CPUSIUvalue=1)이 없는 그룹을 앞 그룹에 합친다. 증설은 장비
        한 대에 대한 변경이라 UPGRADE/DISCO/NEW 가 한 장에 나온다.
      - 장비 이름을 BASE/PROPOSED 구성에서 가져온다.

    신규 견적에서는 합치지 않는다. TS4300 골든의 'No CPUSIU for the following
    products' 그룹은 본체 라인이 없어도 제 장을 갖는다.

    통합 모드는 여기서 두 가지가 갈린다 (근거는 `integrated.py`).
      - 장비 이름을 ProductName 에서 딴다.
      - 본체 LP 에 이미 들어 있는 소프트웨어·서비스 금액을 비운다.
      - 시트명의 금칙 문자를 걷어 내고 겹치는 이름을 갈라 준다.
    """
    is_upgrade = bool(reference_names)
    is_integrated = mode == modes.INTEGRATED

    ordered: list[str] = []
    buckets: dict[str, list[LineItem]] = {}
    for it in items:
        key = it.group_id or it.line_number
        if key not in buckets:
            buckets[key] = []
            ordered.append(key)
        buckets[key].append(it)

    merged: list[str] = []
    for gid in ordered:
        has_body = any(i.siu == 1 for i in buckets[gid])
        if is_upgrade and merged and not has_body:
            buckets[merged[-1]].extend(buckets[gid])
        else:
            merged.append(gid)

    keys: list[str] = []
    members_by_group: list[tuple[LineItem, ...]] = []
    for index, gid in enumerate(merged):
        members = tuple(buckets[gid])
        if is_integrated:
            members = integrated.fold_prices(members)
            key = integrated.group_key(members)
        elif index < len(reference_names):
            key = reference_names[index]
        else:
            key = item_key(members[0].description)
        keys.append(key)
        members_by_group.append(members)

    if is_integrated:
        titles = [sheet_name(k) for k in keys]
        names = unique_sheet_names([safe_sheet_name(k) for k in keys])
    else:
        titles = [""] * len(keys)
        names = [sheet_name(k) for k in keys]

    return tuple(
        Group(group_id=gid, item_key=key, sheet_name=name, title=title,
              items=members)
        for gid, key, name, title, members
        in zip(merged, keys, names, titles, members_by_group)
    )


def _parser() -> etree.XMLParser:
    """외부 참조를 하지 않는 파서.

    인라인 DTD 가 있는 2005년 형식 문서도 읽어야 하므로 문서를 거부하지는 않되,
    엔티티 확장과 네트워크 접근은 막는다.
    """
    return etree.XMLParser(
        load_dtd=False,          # 인라인 DTD 를 읽되 외부 참조는 하지 않는다
        resolve_entities=False,  # XXE 차단
        no_network=True,
        dtd_validation=False,
        huge_tree=False,         # 깊이·크기 폭주 문서 거부
        recover=False,
    )


#: XML 선언의 encoding 값. 선언은 규격상 문서 맨 앞에만 올 수 있다.
_DECLARED_ENCODING = re.compile(
    r'\A(<\?xml[^>]*?encoding\s*=\s*["\'])([A-Za-z0-9_.:+-]+)(["\'])')


def _utf8_equivalent(raw: bytes) -> bytes | None:
    """libxml2 가 모르는 인코딩의 문서를 뜻이 같은 UTF-8 문서로 옮긴다.

    libxml2 는 빌드에 iconv 가 없으면 EUC-KR 같은 인코딩을 거부한다. Pyodide
    (Cloudflare Python Worker, 브라우저 엔진)의 libxml2 가 그렇다.

        XMLSyntaxError: Unsupported encoding EUC-KR, line 1, column 38

    파이썬 표준 코덱은 그 인코딩들을 모두 안다. 그래서 **파싱이 아예 안 되는
    경우에 한해** 파이썬으로 디코딩해 UTF-8 로 다시 적는다. 문자는 하나도
    바뀌지 않으므로 파서가 보는 문서는 데스크톱(iconv 있는 libxml2)이 보는
    것과 같다. 이미 읽히는 문서에는 이 경로가 닿지 않는다.

    Returns:
        옮긴 바이트. 선언이 없거나 파이썬도 모르는 인코딩이면 None.
    """
    head = raw[:512].decode("ascii", "replace")
    match = _DECLARED_ENCODING.match(head)
    if not match:
        return None

    name = match.group(2)
    if name.lower().replace("-", "_") in ("utf_8", "utf8", "us_ascii", "ascii"):
        return None  # libxml2 가 이미 아는 인코딩이라면 다른 이유로 실패한 것이다
    try:
        codec = codecs.lookup(name)
    except LookupError:
        return None

    try:
        text = codec.decode(raw)[0]
    except (UnicodeDecodeError, ValueError):
        return None

    # 선언만 UTF-8 로 고친다. 본문은 손대지 않는다.
    return _DECLARED_ENCODING.sub(
        lambda m: f'{m.group(1)}UTF-8{m.group(3)}', text, count=1).encode("utf-8")


def parse(source: str | Path, *, mode: str = modes.DEFAULT) -> Quotation:
    """eConfig XML 파일 -> Quotation. (데스크톱 경로 입력)

    Args:
        mode: 문서를 읽는 방식. `modes.UNIX` 또는 `modes.INTEGRATED`.

    Raises:
        QuotationXmlError: 로드 실패 또는 필수 노드 누락.
    """
    try:
        tree = etree.parse(str(source), _parser())
    except etree.XMLSyntaxError as exc:
        try:
            raw = Path(source).read_bytes()
        except OSError:
            raw = b""  # 다시 읽을 수 없으면 처음 오류 그대로 알린다
        return _build(_retry_as_utf8(raw, exc), mode)
    return _build(tree.getroot(), mode)


def _retry_as_utf8(raw: bytes, original: "etree.XMLSyntaxError"):
    """인코딩 때문에 막힌 것이면 UTF-8 로 옮겨 한 번만 다시 읽는다.

    다시 읽어도 안 되면 **처음 오류 그대로** 알린다. 사용자가 보는 문구는
    원본 프로그램과 같아야 하고, 대체 경로가 있다는 사실이 드러나면 안 된다.
    """
    retry = _utf8_equivalent(raw)
    if retry is not None:
        try:
            return etree.fromstring(retry, _parser())
        except etree.XMLSyntaxError:
            pass
    raise QuotationXmlError(
        f"XML을 로드하는중 장애 발생. 장애코드: {original}") from original


def parse_bytes(data: bytes, *, mode: str = modes.DEFAULT) -> Quotation:
    """eConfig XML 바이트 -> Quotation. (웹 업로드 입력)

    파일시스템을 건드리지 않는다. 인코딩은 XML 선언을 따르므로 UTF-8 과
    EUC-KR 문서를 모두 그대로 받는다. 문자열이 아니라 **바이트** 를 넘겨야
    선언된 인코딩이 존중된다.

    Raises:
        QuotationXmlError: 로드 실패 또는 필수 노드 누락.
    """
    raw = bytes(data)
    if not raw.strip():
        raise QuotationXmlError("XML을 로드하는중 장애 발생. 장애코드: 빈 화일입니다.")
    try:
        root = etree.fromstring(raw, _parser())
    except etree.XMLSyntaxError as exc:
        root = _retry_as_utf8(raw, exc)
    return _build(root, mode)


def _build(root, mode: str = modes.DEFAULT) -> Quotation:
    """파싱된 문서 뿌리 -> Quotation. 경로 입력과 바이트 입력의 공통 경로다."""
    if etree.QName(root).localname != "CFXML":
        raise QuotationXmlError("CFXML을 찾을수 없습니다.")

    cfdata = root.find(XP_CFDATA)
    if cfdata is None:
        raise QuotationXmlError("CFData을 찾을수 없습니다.")

    elements = cfdata.findall(XP_LINE_ITEM)
    if not elements:
        raise QuotationXmlError("견적서 작성을 위한 Item을 찾을 수 없습니다.")

    all_items = [_parse_line(e) for e in elements]
    # 증설 견적은 기존(BASE)·증설후(PROPOSED) 구성을 참조용으로 함께 담는다.
    # 견적서에는 실제 증설분만 넣는다.
    quoted = [i for i in all_items if not i.is_reference]
    if not quoted:
        raise QuotationXmlError("견적서 작성을 위한 Item을 찾을 수 없습니다.")

    return Quotation(
        groups=_build_groups(quoted, _reference_names(all_items), mode))

File: rust/core/src/xml_reader.rs
//! eConfig XML 파서. 파이썬 `quotation/core/xml_reader.py`.
//!
//! 문서를 읽는 **규칙**은 파이썬과 같다. 파서만 lxml(libxml2)에서
//! `quick-xml` 로 바뀌었고, 그 차이는 [`crate::xmldom`] 에 모아 두었다.
//!
//! 파이썬 쪽 EUC-KR 우회책(`_utf8_equivalent`)에 해당하는 코드는 없다.
//! `encoding_rs` 가 EUC-KR 을 기본으로 알기 때문에 선언만 보면 된다. 우회책이
//! 왜 있었는지는 [사고 0002](../../../doc/incidents/0002-pyodide-lxml-euckr.md)
//! 에 남아 있다.
//!
//! 한 곳이 다르다. 파이썬은 수량이나 금액이 숫자가 아니면 `ValueError` /
//! `InvalidOperation` 으로 죽는다. 여기서는 같은 자리에서 문서 오류로 알린다.
//! 통과하던 문서의 결과는 달라지지 않고, 죽던 자리에서만 문구가 생긴다.

use crate::dcsc_summary::{self, Summary};
use crate::integrated;
use crate::models::{Group, LineItem, Quotation, SubLineItem};
use crate::modes::{self, Mode};
use crate::money;
use crate::naming::{item_key, safe_sheet_name, sheet_name, unique_sheet_names};
use crate::xmldom::{self, Element};

/// 상위 라인과 서브라인을 합친 문서 전체 품목 수 상한.
pub const MAX_QUOTATION_ITEMS: usize = 5_000;

const LOAD_FAILURE: &str = "XML을 로드하는중 장애 발생. 장애코드: ";
const NO_ITEMS: &str = "견적서 작성을 위한 Item을 찾을 수 없습니다.";

/// XML 이 견적서 생성 요건을 만족하지 않을 때. 문구는 원본 프로그램과 같다.
#[derive(Debug, PartialEq, Eq)]
pub struct QuotationXmlError {
    pub message: String,
}

impl QuotationXmlError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }

    fn load(detail: impl std::fmt::Display) -> Self {
        Self::new(format!("{LOAD_FAILURE}{detail}"))
    }
}

impl std::fmt::Display for QuotationXmlError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}

impl std::error::Error for QuotationXmlError {}

type Result<T> = std::result::Result<T, QuotationXmlError>;

// --- 원본 프로그램의 경로 (변경 금지) -----------------------------------------

const P_CFDATA: [&str; 1] = ["CFData"];
const P_LINE_NUMBER: [&str; 1] = ["ProductLineNumber"];
const P_SIU: [&str; 1] = ["CPUSIUvalue"];
const P_TXN_TYPE: [&str; 1] = ["TransactionType"];
const P_GROUP_ID: [&str; 1] = ["ProprietaryGroupIdentifier"];
const P_QUANTITY: [&str; 1] = ["Quantity"];
const P_DESC: [&str; 3] = [
    "ProductIdentification",
    "PartnerProductIdentification",
    "ProductDescription",
];
const P_PRODUCT_NAME: [&str; 3] = [
    "ProductIdentification",
    "PartnerProductIdentification",
    "ProductName",
];
const P_TYPE_CODE: [&str; 3] = [
    "ProductIdentification",
    "PartnerProductIdentification",
    "ProductTypeCode",
];
const P_PART_NO: [&str; 3] = [
    "ProductIdentification",
    "PartnerProductIdentification",
    "ProprietaryProductIdentifier",
];
const P_AMOUNT: [&str; 3] = ["UnitListPrice", "FinancialAmount", "MonetaryAmount"];
const LINE_ITEM: &str = "ProductLineItem";
const SUB_LINE_ITEM: &str = "ProductSubLineItem";

/// eConfig XML 바이트 -> 견적서 한 부.
///
/// `mode` 를 주면 문서를 읽는 방식을 강제한다 (`"unix"` / `"integrated"`).
/// 비워 두면 문서 내용으로 알아낸다 — 보통은 이쪽을 쓴다.
pub fn parse_bytes(data: &[u8], mode: Option<&str>) -> Result<Quotation> {
    let (root, items) = read(data)?;

    // 증설 견적은 기존(BASE)·증설후(PROPOSED) 구성을 참조용으로 함께 담는다.
    // 견적서에는 실제 증설분만 넣는다.
    let names = reference_names(&items);
    let quoted: Vec<LineItem> = items
        .into_iter()
        .filter(|item| !item.is_reference())
        .collect();
    if quoted.is_empty() {
        return Err(QuotationXmlError::new(NO_ITEMS));
    }

    // IBM 문서인지 레노버 x86 문서인지는 사람이 고르지 않는다. 본체 라인에
    // ProductName 이 있으면 통합, 없으면 UNIX 다.
    let resolved = modes::resolve(mode, quoted.iter().map(|item| item.product_name.as_str()))
        .map_err(|error| QuotationXmlError::new(error.to_string()))?;

    // 요약표는 문서 하나에 한 벌이고 장비군에 한 번씩 짝지어 쓴다. 통합 모드가
    // 아니면 읽지 않는다 — UNIX 경로는 한 줄도 달라지지 않는다.
    let mut summary = match resolved {
        Mode::Integrated => Some(dcsc_summary::parse(&root)),
        Mode::Unix => None,
    };

    Ok(Quotation {
        mode: resolved,
        groups: build_groups(quoted, &names, resolved, summary.as_mut()),
    })
}

/// eConfig XML 바이트 -> 문서에 적힌 모든 라인 (참조용 구성까지 그대로).
pub fn parse_items(data: &[u8]) -> Result<Vec<LineItem>> {
    let (_, items) = read(data)?;
    if items.iter().all(LineItem::is_reference) {
        return Err(QuotationXmlError::new(NO_ITEMS));
    }
    Ok(items)
}

/// 문서를 읽어 뿌리와 라인을 함께 돌려준다. 요약표가 뿌리를 다시 본다.
fn read(data: &[u8]) -> Result<(Element, Vec<LineItem>)> {
    if data.iter().all(u8::is_ascii_whitespace) {
        return Err(QuotationXmlError::load("빈 화일입니다."));
    }

    let root = xmldom::parse(data).map_err(QuotationXmlError::load)?;
    if root.name != "CFXML" {
        return Err(QuotationXmlError::new("CFXML을 찾을수 없습니다."));
    }
    let cfdata = root
        .find(&P_CFDATA)
        .ok_or_else(|| QuotationXmlError::new("CFData을 찾을수 없습니다."))?;

    let mut elements = Vec::new();
    cfdata.descendants_named(LINE_ITEM, &mut elements);
    if elements.is_empty() {
        return Err(QuotationXmlError::new(NO_ITEMS));
    }

    let count = elements.len()
        + elements
            .iter()
            .map(|element| element.children_named(SUB_LINE_ITEM).count())
            .sum::<usize>();
    if count > MAX_QUOTATION_ITEMS {
        return Err(QuotationXmlError::new(format!(
            "구성 품목이 너무 많습니다. (최대 {}건)",
            with_thousands(MAX_QUOTATION_ITEMS)
        )));
    }

    let items = elements
        .iter()
        .map(|element| read_line(element))
        .collect::<Result<Vec<_>>>()?;
    Ok((root, items))
}

fn read_line(element: &Element) -> Result<LineItem> {
    let subs = element
        .children_named(SUB_LINE_ITEM)
        .map(read_sub)
        .collect::<Result<Vec<_>>>()?;
    Ok(LineItem {
        line_number: element.text_at(&P_LINE_NUMBER).to_owned(),
        txn_type: element.text_at(&P_TXN_TYPE).to_owned(),
        group_id: element.text_at(&P_GROUP_ID).to_owned(),
        quantity: int_at(element, &P_QUANTITY, 1)?,
        part_number: element.text_at(&P_PART_NO).to_owned(),
        description: element.text_at(&P_DESC).to_owned(),
        product_type: element.text_at(&P_TYPE_CODE).to_owned(),
        product_name: element.text_at(&P_PRODUCT_NAME).to_owned(),
        unit_price: amount_at(element, &P_AMOUNT)?,
        subs,
        siu: int_at(element, &P_SIU, 0)?,
    })
}

fn read_sub(element: &Element) -> Result<SubLineItem> {
    Ok(SubLineItem {
        txn_type: element.text_at(&P_TXN_TYPE).to_owned(),
        quantity: int_at(element, &P_QUANTITY, 1)?,
        part_number: element.text_at(&P_PART_NO).to_owned(),
        description: element.text_at(&P_DESC).to_owned(),
        unit_price: amount_at(element, &P_AMOUNT)?,
    })
}

/// 파이썬 `_int` — 비어 있으면 기본값.
fn int_at(element: &Element, path: &[&str], default: i64) -> Result<i64> {
    let raw = element.text_at(path);
    if raw.is_empty() {
        return Ok(default);
    }
    raw.parse::<i64>()
        .map_err(|_| QuotationXmlError::load(format!("숫자가 아닙니다: {raw}")))
}

fn amount_at(element: &Element, path: &[&str]) -> Result<money::Amount> {
    money::parse_amount(Some(element.text_at(path)))
        .map_err(|error| QuotationXmlError::load(error.to_string()))
}

/// 증설 견적의 장비 이름. 기존(BASE)/증설후(PROPOSED) 구성의 본체 라인에서 딴다.
///
/// 증설 라인(UPGRADE)의 Description 은 `9080 Model HEU` 처럼 장비 이름이 없다.
/// 골든은 시트를 `SERVER 1` 로 부르는데, 그 이름은 BASE 구성의 본체 라인
/// `Server 1:Server 1:IBM Power E1080` 에서 온다.
fn reference_names(items: &[LineItem]) -> Vec<String> {
    let mut names: Vec<String> = Vec::new();
    for item in items {
        if item.is_reference() && item.siu == 1 {
            let key = item_key(&item.description);
            if !names.contains(&key) {
                names.push(key);
            }
        }
    }
    names
}

/// ProprietaryGroupIdentifier 로 묶는다. 문서 등장 순서를 유지한다.
///
/// 증설 견적일 때만 두 가지 보정이 붙는다.
///   - 본체 라인(CPUSIUvalue=1)이 없는 그룹을 앞 그룹에 합친다. 증설은 장비
///     한 대에 대한 변경이라 UPGRADE/DISCO/NEW 가 한 장에 나온다.
///   - 장비 이름을 BASE/PROPOSED 구성에서 가져온다.
///
/// 신규 견적에서는 합치지 않는다. 본체 라인이 없어도 제 장을 갖는다.
///
/// 통합 모드는 장비 이름을 ProductName 에서 따고, 본체 LP 에 들어 있는
/// 소프트웨어·서비스 금액을 갈라 놓으며, 시트명의 금칙 문자를 걷어 낸다.
fn build_groups(
    items: Vec<LineItem>,
    reference_names: &[String],
    mode: Mode,
    mut summary: Option<&mut Summary>,
) -> Vec<Group> {
    let is_upgrade = !reference_names.is_empty();
    let is_integrated = mode == Mode::Integrated;

    let mut buckets: Vec<(String, Vec<LineItem>)> = Vec::new();
    for item in items {
        let key = if item.group_id.is_empty() {
            item.line_number.clone()
        } else {
            item.group_id.clone()
        };
        match buckets.iter_mut().find(|(id, _)| *id == key) {
            Some((_, members)) => members.push(item),
            None => buckets.push((key, vec![item])),
        }
    }

    let mut merged: Vec<(String, Vec<LineItem>)> = Vec::new();
    for (id, members) in buckets {
        let has_body = members.iter().any(|item| item.siu == 1);
        match merged.last_mut() {
            Some((_, previous)) if is_upgrade && !has_body => previous.extend(members),
            _ => merged.push((id, members)),
        }
    }

    let mut ids: Vec<String> = Vec::with_capacity(merged.len());
    let mut keys: Vec<String> = Vec::with_capacity(merged.len());
    let mut members_by_group: Vec<Vec<LineItem>> = Vec::with_capacity(merged.len());
    for (index, (id, members)) in merged.into_iter().enumerate() {
        let (key, members) = if is_integrated {
            let members = integrated::fold_prices(members, summary.as_deref_mut());
            let key = integrated::group_key(&members);
            (key, members)
        } else if index < reference_names.len() {
            (reference_names[index].clone(), members)
        } else {
            (item_key(&members[0].description), members)
        };
        ids.push(id);
        keys.push(key);
        members_by_group.push(members);
    }

    let (titles, names): (Vec<String>, Vec<String>) = if is_integrated {
        (
            keys.iter().map(|key| sheet_name(key)).collect(),
            unique_sheet_names(keys.iter().map(|key| safe_sheet_name(key))),
        )
    } else {
        (
            vec![String::new(); keys.len()],
            keys.iter().map(|key| sheet_name(key)).collect(),
        )
    };

    ids.into_iter()
        .zip(keys)
        .zip(names)
        .zip(titles)
        .zip(members_by_group)
        .map(
            |((((group_id, item_key), sheet_name), title), items)| Group {
                group_id,
                item_key,
                sheet_name,
                title,
                items,
            },
        )
        .collect()
}

/// 파이썬 `f"{n:,}"`.
fn with_thousands(value: usize) -> String {
    let digits = value.to_string();
    let mut out = String::with_capacity(digits.len() + digits.len() / 3);
    for (index, digit) in digits.chars().enumerate() {
        if index > 0 && (digits.len() - index) % 3 == 0 {
            out.push(',');
        }
        out.push(digit);
    }
    out
}
